### DISCUSSION TITLE
How to use ORM table foreign key constraints on Selectable/Subquery object?

### DISCUSSION THREAD

**Discussion by jlynchMicron** [sqlalchemy/sqlalchemy#9479] [Usage Questions]
I have to do a pre-filtering query on a table before using the results to join against another table, how do you maintain the foreign key constraints on the original table through to the next join statement?

In my case, I am using the BigQuery dialect and I need to do an "unnest" function on a table before I use the results to join against another table, where both tables have a pre-defined ORM foreign key constraint.  This constraint does not seem to be manifesting in the WHERE clause of my query, like it would if I did a natural join against the two tables.

TLDR:
How do I maintain foreign key constraints through Selectable/Subquery objects?

Pseudocode Example:
```
stmt_1 = sa.select( [tbl1_pkeys] + sa.func.unnest( tbl1.c['array'] ) ).select_from(tbl1)
stmt_2 = sa.select('*').select_from(stmt_1.join(tbl2))
```

**Comment by CaselIT** [comment]
Hi,

I'm guessing you are looking for a way other than explicitly writing the join condition, right?

**Reply by jlynchMicron** [reply]
Correct.  Since the stmt_1 query is still only operating on tbl1, I feel like it should carry through its foreign key constraints to the tbl2 join.  I think pre-filtering queries are not unusual when working with big data to reduce join computation to only what is needed.

In my specific case, I want to pre-filter on a sub-element of the column I am unnesting, before performing the join.  In my exact problem, my column is an Array of Records, and the only way to filter on it is if its unnested first.

If its not possible to carry through the table foreign key constraints through on the Selectable, is there a way I can manually insert the table's foreign key constraints into the where condition of the join?  I need this to be programmatically performed by the constraints in my program since my code logic is used for multiple different tables.

**Reply by CaselIT** [reply]
Note that this should work already. If you want further help you will need to provide a runnable example that reproduces the issue.
Also indicate the sqlachemy version in use

**Reply by jlynchMicron** [reply]
I can try and provide an example in a bit, but do you think this could be caused by the unnest operation, which gives the Selectable 2 items in its "froms" variable?  I am using SQLalchemy version 1.4

![image](https://user-images.githubusercontent.com/58755604/225071237-8150fada-70e0-46ce-bcb7-b84d35836029.png)

**Reply by CaselIT** [reply]
trying this on a simple query seems to work even with unnest in the inner select.

**Reply by jlynchMicron** [reply]
Oh man, I got too caught up in the query and I think this issue is due to me specifying additional where conditions in the join that are overwriting the foreign key constraints. When I remove the extra join conditions, the query resorts to a natural join based on foreign keys.

Is there a way to include both the foreign key constraints and additional join columns in a join command?  I was assuming the fkeys would be applied to the join implicitly, but that does not seem to be the case when explicitly expressing extra join conditions.

Example:
```
stmt_1.join(tbl2, natural join fkeys + extra manual conditions)
```

In my case, the extra manual conditions will come from my unnested column elements.  Sorry for being cryptic about my query, just trying to avoid having my company complain about security concerns.

If this is not possible, I guess I could just add the filtering as a WHERE statement on the join or something.

**Reply by CaselIT** [reply]
I'm not sure. If the query is an inner join you can just place them to the where.
@zzzeek do you have idea here?

Otherwise you can create a second subquery, after the unnest and filter stuff there. Something like
```py
q1 = sa.select(tbl1, unnest(...).label('my_col')).where(...).subquery('inner')
q2 = sa.select(q1).where(q1.c.my_col > 2).subquery('outer')
q3 = sa.select(tbl2, q2).select_from(join(tbl2, q2))
```
